### PR TITLE
Finally, diff-cover!

### PR DESCRIPTION
**Pull request by bocajnotnef**
I'm excited.

I've added a coverage detecting agent that will parse a git diff and check the changes you made against available coverage metrics. If the changes you made aren't covered by tests, you get yelled at.

The script requires python3, and is invoked with `make diff-cover` in the project root.

The script will ignore files listed in `scripts/ignore_list`, which is useful for making it ignore the test driver files themselves.

Its output looks like this:

```
python3 dumb_diff_cover.py -i ignore_list diff
Warning: removing file tests/test_driver.cc from concern
Warning: removing file tests/test_tools.cc from concern
Files to check coverage: 
        geometry/Angle2D.h
        tools/sequence_utils.h
        tools/assert.h
        tools/vector.h
        tests/test_scholar.cc
        geometry/Circle2D.h
        tools/fixed.h
        tools/BitMatrix.h
        tools/unit_tests.h
        tools/FunctionSet.h
        scholar/Author.h
        tools/ProbSchedule.h
        tools/bitset_utils.h
        tests/test_geometry.cc
        tools/BitSet.h
        tools/const.h
        tools/mem_track.h
        tools/string_utils.h
        tools/const_utils.h
        tools/tuple_struct.h
        tools/BitVector.h
        tools/graph_utils.h
        geometry/Point2D.h
        tools/serialize_macros.h
        tools/Trait.h
        tools/alert.h
        tools/DynamicStringSet.h
        tools/Graph.h
        tools/serialize.h
        tools/macro_math.h
        tools/macros.h
        tools/reflection.h
        tools/Ptr.h
        tools/Random.h
        tools/functions.h
        tools/class.h
        tools/random_utils.h
        tools/errors.h
        third-party/catch/single_include/catch.hpp
        tools/SolveState.h
        geometry/Body2D.h

Files that aren't covered at all, but are changed: 
        evo/NK.h


Checking coverage of tools/BitMatrix.h
        Bad lines for tools/BitMatrix.h

                #####:   93:    void SetAll() { bits.SetAll(); }
                #####:   94:    void SetCol(int col) { bits |= MaskCol<0>() << col;}
                #####:   95:    void SetRow(int row) { bits |= (MaskRow<0>() << (row * COLS)); }
                #####:   96:    void Clear() { bits.Clear(); }
                #####:   97:    void ClearCol(int col) { bits &= ~(MaskCol<0>() << col); }
                #####:   98:    void ClearRow(int row) { bits &= ~(MaskRow<0>() << (row * COLS)); }
                #####:  111:    BitMatrix UpShift() const { return bits >> COLS; }
                #####:  132:        int potato = 7;


Checking coverage of tools/BitSet.h
        Bad lines for tools/BitSet.h

                #####:  292:    void flip(int start, int end) {
                #####:  293:      assert(start >= 0 && start <= end && end <= NUM_BITS);
                #####:  295:        Set(index, ~Get(index));
                #####:  302:    void flip(int index) {
                #####:  303:      Set(index, ~Get(index));


Checking coverage of tools/Random.h
        Bad lines for tools/Random.h

                -:  324:    int pheasant = 7;


Checking coverage of tools/BitVector.h
** Found uncovered lines, failing!
```

which I think provides sufficient information to fix any coverage holes you introduce.

I'd like to have this integrated into the contribution process--If we enforce the use of this it'll help a lot to get solid coverage across the library, and is a damn good stopgap measure until we find an 'official' way to accurately metric test coverage across the repo.

Along the way I also rewrote the dependency mapper so that it didn't suck. Turns out knowing how graphs work is useful when building a dependency tree--who knew? I also updated the `tools/LEVEL_MAP` to be correct.

### CODE REVIEW

**Comment by bocajnotnef** [comment]
Looks like Travis is gonna break on this because the g++ compiler and the gcov version on Travis are mismatched, so its segfaulting.

Always something, eh?

@stevenjson @emilydolson Review & test, please?

**Comment by mercere99** [comment]
Should we pull the "Ready For Merge" label until the Travis issue is fixed?

Otherwise, looks like a fantastic feature!

**Comment by bocajnotnef** [comment]
Travis issue resolved by #79, though with some cost (long build times).

Also added tests so diff-cover can stop failing its own branch (however amusing that is)

**Comment by stevenjson** [comment]
dumb_diff_cover.py isn't really professional sounding

**Comment by stevenjson** [comment]
You should make it so the script tells you the Pathname to the file that is uncovered

**Comment by bocajnotnef** [comment]
@stevenjson So I was working on getting the path names to show up and was having trouble reproducing the problem we saw when testing (i.e. there not being a path name), 'cuz for me it was showing up like:

image

Then I remembered, the way the git diff is structured the filename is listed like:
`diff --git a/.gitignore b/.gitignore`, and I was splitting the string based off of `b/`.

The directory you used to test was `bb/`.

The split broke it.

So, as long as the filepath doesn't end in 'b' the script should properly grab the path....

**Comment by bocajnotnef** [comment]
Y'know, hypothetically speaking, it wouldn't be the most difficult thing in the world to add functionality to this script that would render "correct" test coverage for the whole repo, in pretty pretty html...

Might be worth considering future work in that direction.

**Comment by bocajnotnef** [comment]
Fun thing I just noticed:

If someone, say @mercere99, adds code to master and I don't pull the code into my branch, and I run vigilance, @mercere99's code gets detected as an 'addition' and (in this case) wasn't covered by tests, so vigilance yelled at me.

tl;dr: `git fetch --all && git pull origin master` prior to running vigilance/`diff-cover`, or you'll get yelled at for other people's changes.
